examples/bidspath.py

```diff
@@ -0,0 +1,159 @@
+# also be used when creating a ``BIDSPath``.
+#
+# Many studies consist of multiple **sessions**. As you may have guessed,
+# BIDS specifies how to store data for each session, and consequently,
+# ``BIDSPath`` handles this for you too! Let's update our ``BIDSPath`` with
+# information on our experimental session, and try to retrieve it again via
+# ``.session``.
+
+session = 'test'
+bids_path.update(session=session)
+print(bids_path.session)
+
+###############################################################################
+# Now that was easy! We're almost there! We also need to specify a
+# **data type**, i.e., ``meg`` for MEG data, ``eeg`` and ``ieeg`` for EEG and
+# iEEG data, or ``anat`` for anatomical MRI scans. Typically, MNE-BIDS will
+# infer the data type of your data automatically, for example when writing data
+# using `mne_bids.write_raw_bids`. For the sake of this example, however, we
+# are going to specify the data type explicitly.
+
+datatype = 'meg'
+bids_path.update(datatype=datatype)
+print(bids_path.datatype)
+
+###############################################################################
+# Excellent! Let's have a look at the path we have constructed!
+print(bids_path)
+
+###############################################################################
+# As you can see, ``BIDSPath`` automatically arranged all the information we
+# provided such that it creates a valid BIDS folder structure. You can also
+# retrieve a `pathlib.Path` object of this path:
+
+pathlib_path = bids_path.fpath
+pathlib_path
+
+###############################################################################
+# Let's have a closer look at the components of our ``BIDSPath`` again.
+
+bids_path
+
+###############################################################################
+# The most interesting thing here is probably the **basename**. It's what
+# MNE-BIDS uses to name individual files. The basename consists of a set of
+# so-called **entities**, which are concatenated using underscores. You can
+# access it directly:
+
+bids_path.basename
+
+###############################################################################
+# The two entities you can see here are the ``subject`` entity (``sub``) and
+# the ``session`` entity (``ses``). Each entity name also has a value; for
+# ``sub``, this is ``123``, and for ``ses``, it is ``test`` in our example.
+# Entity names (or "keys") and values are separated via hyphens.
+# BIDS knows a much larger number of entities, and MNE-BIDS allows you to make
+# use of them. To get a list of all supported entities, use:
+
+bids_path.entities
+
+###############################################################################
+# As you can see, most entity keys are set to ``None``, which is the default
+# and implies that no value has been set. Let us add a ``run`` entity, and
+# remove the ``session``:
+
+run = '01'
+session = None
+bids_path.update(run=run, session=session)
+bids_path
```

Comment:
print(bids_path) ?

to avoid confusing the user with two different ways to do similar things

Reply:
The thing is that `print(bids_path)` produces a different output than just `bids_path`, because `BIDSPath.__str__` and `BIDSPath.__repr__` are different. But you're right, it might be confusing

Reply:
the `str` essentially calls `fpath`, whereas `repr` is more like Pathlib looking. IIRC I think functionally this made it a lot easier to pass around bids_path to functions that might require a string (e.g. some `os` functions).

I think that's what people defaulted to when reviewing the initial BIDSPath.

Reply:
Yes I remember this discussion too. Just to recap for all of you, this is what it currently looks like:

```python
In [5]: p = BIDSPath(root='/tmp/foo', subject='01', session='test',
   ...:              task='dostuff', datatype='meg')

In [6]: p
Out[6]:
BIDSPath(
root: /tmp/foo
datatype: meg
basename: sub-01_ses-test_task-dostuff)

In [7]: print(p)
/Users/hoechenberger/Development/mne-bids/mne_bids/path.py:445: RuntimeWarning: Could not locate a data file of a supported format. This is likely a problem with your BIDS dataset. Please run the BIDS validator on your data. (root=/tmp/foo, basename=sub-01_ses-test_task-dostuff). []
  warn(msg)
/tmp/foo/sub-01/ses-test/meg/sub-01_ses-test_task-dostuff
```
I'm not so sure about this warning either…
**Plus** the above lines create a new folder structure `/tmp/foo/sub-01/ses-test/meg`, which is probably not what we want either? (`/tmp/foo` did not exist before)

cc @agramfort

Reply:
No warning when I specify suffix and extension:

```python
In [14]: p = BIDSPath(root='/tmp/foo', subject='01', session='test',
    ...:              task='dostuff', datatype='meg', suffix='meg',
    ...:              extension='.fif')

In [15]: print(p)
/tmp/foo/sub-01/ses-test/meg/sub-01_ses-test_task-dostuff_meg.fif
```

Reply:
But generally I like the `repr`. And the `str` being path-like makes it easier to pass a `BIDSPath` to other functions, as @adam2392 pointed out. I will see if and how I can adjust the example without us having to change `BIDSPath`.

Reply:
yes str is to be able to pass a BIDSPath as a regular path in our function. It makes it's path-like. The repr is just for user facing information so it's ok that repr and str are different.

Reply:
okay, makes sense!

Reply:
I'm addressing this in b50b260, check out example 01.